src/Pos/Context/Context.hs

```diff
@@ -36,169 +35,225 @@ import           Universum
+    -- ^ Data necessary for 'MonadSlotsData'.
+    , ncSlottingContext     :: !SlottingContextSum
+    -- ^ Context needed for Slotting.
+    , ncShutdownContext     :: !ShutdownContext
+    -- ^ Context needed for Shutdown
+    , ncBlkSemaphore        :: !BlkSemaphore
+    -- ^ Semaphore which manages access to block application.
+    -- Stored hash is a hash of last applied block.
+    , ncUserSecret          :: !(TVar UserSecret)
+    -- ^ Secret keys (and path to file) which are used to send transactions
+    , ncBlockRetrievalQueue :: !(BlockRetrievalQueue ssc)
+    -- ^ Concurrent queue that holds block headers that are to be
+    -- downloaded.
+    , ncRecoveryHeader      :: !(RecoveryHeader ssc)
+    -- ^ In case of recovery mode this variable holds the latest header hash
+    -- we know about, and the node we're talking to, so we can do chained
+    -- block requests. Invariant: this mvar is full iff we're more than
+    -- 'recoveryHeadersMessage' blocks deep relatively to some valid header
+    -- and we're downloading blocks. Every time we get block that's more
+    -- difficult than this one, we overwrite. Every time we process some
+    -- blocks and fail or see that we've downloaded this header, we clean
+    -- mvar.
+    , ncProgressHeader      :: !(ProgressHeader ssc)
+    -- ^ Header of the last block that was downloaded in retrieving
+    -- queue. Is needed to show smooth prorgess on the frontend.
+    , ncInvPropagationQueue :: !RelayPropagationQueue
+    -- ^ Queue is used in Relay framework,
+    -- it stores inv messages for earlier received data.
+    , ncLoggerConfig        :: !LoggerConfig
+    -- ^ Logger config, as taken/read from CLI.
+    , ncNodeParams          :: !NodeParams
+    -- ^ Params node is launched with
+    , ncStartTime           :: !StartTime
+    -- ^ Time when node was started ('NodeContext' initialized).
+    , ncLastKnownHeader     :: !(LastKnownHeader ssc)
+    -- ^ Header of last known block, generated by network (announcement of
+    -- which reached us). Should be use only for informational purposes
+    -- (status in Daedalus). It's easy to falsify this value.
+    , ncTxpGlobalSettings   :: !TxpGlobalSettings
+    -- ^ Settings for global Txp.
+    , ncConnectedPeers      :: !ConnectedPeers
+    -- ^ Set of peers that we're connected to.
+    }
+
+makeLensesWith postfixLFields ''NodeContext
+
+class HasNodeContext ssc ctx | ctx -> ssc where
+    nodeContext :: Lens' ctx (NodeContext ssc)
+
+instance HasNodeContext ssc (NodeContext ssc) where
+    nodeContext = identity
+
+instance HasSscContext ssc (NodeContext ssc) where
+    sscContext = ncSscContext_L
+
+instance HasDiscoveryContextSum (NodeContext ssc) where
+    discoveryContextSum = ncDiscoveryContext_L
+
+instance HasSlottingVar (NodeContext ssc) where
+    slottingTimestamp = ncSlottingVar_L . _1
+    slottingVar = ncSlottingVar_L . _2
+
+instance HasLens SlottingContextSum (NodeContext ssc) SlottingContextSum where
```

Comment:
After reviewing majority of this PR I suddenly decided to ask: why do you use `HasPatak` in some places and `HasLens Bardaq` in other places? Is it WIP or planned for another task or http://cs5.pikabu.ru/post_img/big/2015/12/04/5/1449210847155432089.jpg?